### PR TITLE
Beautify Shade labels 

### PR DESCRIPTION
**Pull request by daschuer**

### CODE REVIEW

File: src/widget/woverview.cpp [hunk 1]
```diff
@@ -402,8 +402,8 @@ void WOverview::mouseMoveEvent(QMouseEvent* e) {
     // the hotcue rendered on top must be assigned to m_pHoveredMark to show
     // the CueMenu. To accomplish this, m_marksToRender is iterated in reverse
     // and the loop breaks as soon as m_pHoveredMark is set.
-    for (auto it = m_marksToRender.crbegin(); it != m_marksToRender.crend(); ++it) {
-        auto pMark = *it;
+    for (int i = m_marksToRender.size() - 1; i >= 0 ; --i) {
```
Comment (Be-ing) [inline src/widget/woverview.cpp:405]: Thanks, this is easier to understand anyway.

File: res/skins/Shade/deck_overview.xml [hunk 2]
```diff
@@ -14,9 +14,6 @@
     <PlayPosColor>#00FF00</PlayPosColor>
     <EndOfTrackColor>#EA0000</EndOfTrackColor>
     <PlayedOverlayColor>#60000000</PlayedOverlayColor>
-    <LabelBackgroundColor>#80000000</LabelBackgroundColor>
-    <LabelTextColor>#ffffff</LabelTextColor>
-    <LabelFontSize>9</LabelFontSize>
```
Comment (Be-ing) [inline res/skins/Shade/deck_overview.xml:19]: This breaks Shade, the label disappears when hovering:
image
9 is the maximum font size that fits

File: src/waveform/waveformmarklabel.cpp [hunk 3]
```diff
@@ -47,9 +47,9 @@ void WaveformMarkLabel::prerender(QPointF bottomLeft, QPixmap icon, QString text
 
     QPainter painter(&m_pixmap);
 
-    painter.setPen(backgroundColor);
+    painter.setPen(QColor(Qt::transparent));
     painter.setBrush(QBrush(backgroundColor));
-    painter.drawRoundedRect(0, 0, pixmapRect.width(), pixmapRect.height(), 5.0, 5.0);
+    painter.drawRoundedRect(0, 0, pixmapRect.width(), pixmapRect.height(), 2.0, 2.0);
```
Comment (Be-ing) [inline src/waveform/waveformmarklabel.cpp:52]: Why was this changed?

File: src/waveform/waveformmarklabel.cpp [hunk 4]
```diff
@@ -47,9 +47,9 @@ void WaveformMarkLabel::prerender(QPointF bottomLeft, QPixmap icon, QString text
 
     QPainter painter(&m_pixmap);
 
-    painter.setPen(backgroundColor);
+    painter.setPen(QColor(Qt::transparent));
```
Comment (Be-ing) [inline src/waveform/waveformmarklabel.cpp:50]: Why was this changed?

File: res/skins/Shade/deck_overview.xml (hunk 2 above)
Comment (Be-ing) [reply res/skins/Shade/deck_overview.xml:19]: If you want to increase the vertical height of the overview to allow a bigger text size, that would be fine with me.

Comment (daschuer) [reply res/skins/Shade/deck_overview.xml:19]: Not for me: 
grafik

Does this overlap also happen in a unscaled version? ... Or is it a different font issue?

File: src/waveform/waveformmarklabel.cpp (hunk 3 above)
Comment (daschuer) [reply src/waveform/waveformmarklabel.cpp:52]: Because with 5.0, the labels are too much rounded, almost a circle in case of a single character. 
See: grafik

File: src/waveform/waveformmarklabel.cpp (hunk 4 above)
Comment (daschuer) [reply src/waveform/waveformmarklabel.cpp:50]: To remove the distracting border: 

grafik

File: res/skins/Shade/deck_overview.xml (hunk 2 above)
Comment (Be-ing) [reply res/skins/Shade/deck_overview.xml:19]: This also happens for me with 1x scaling:
image

Comment (daschuer) [reply res/skins/Shade/deck_overview.xml:19]: Ok, so I guess we need to set a font as well.

Comment (daschuer) [reply res/skins/Shade/deck_overview.xml:19]: The attached picture is not 1x scaling. 
If I scale mine, the fonts are both 16 px. 
So the difference is the background size.

Comment (Be-ing) [reply res/skins/Shade/deck_overview.xml:19]: You're right, I still had automatic scaling enabled so that was at 2x. Here it is at 1x; the problem is still there:
image

Comment (daschuer) [reply res/skins/Shade/deck_overview.xml:19]: Got it. See the latest commit. Your font seems to have higher charters for Î or such. But this is not relevant here, so we can use the pixel high, which only encounters the height of "Hg"

File: src/waveform/waveformmarklabel.cpp (hunk 4 above)
Comment (Be-ing) [reply src/waveform/waveformmarklabel.cpp:50]: okay

File: src/waveform/waveformmarklabel.cpp [hunk 5]
```diff
@@ -28,7 +28,11 @@ void WaveformMarkLabel::prerender(QPointF bottomLeft, QPixmap icon, QString text
         text = fontMetrics.elidedText(text, Qt::ElideRight, availableWidthForText);
         pixmapRect.setWidth(widgetWidth);
     }
-    pixmapRect.setHeight(math_max(fontMetrics.height(), icon.height()));
+    // We take the pixelSize for background height, to not consider exotic
+    // high characters.
+    pixmapRect.setHeight(
+            math_max(font.pixelSize() * 1.2,
```
Comment (Be-ing) [inline src/waveform/waveformmarklabel.cpp]: As far as I can tell this makes no difference. The label is still hidden in Shade when I hover a hotcue.

File: res/skins/Shade/deck_overview.xml (hunk 2 above)
Comment (Be-ing) [reply res/skins/Shade/deck_overview.xml:19]: Why should we assume that the user isn't using tall characters like Î? The label can be any text the user wants.

Comment (daschuer) [reply res/skins/Shade/deck_overview.xml:19]: The user can still use Î, depending to the font it may now hit or exceed the background. This is OK for me since the majority of characters are looking well.

Comment (Be-ing) [reply res/skins/Shade/deck_overview.xml:19]: I don't think this is a good idea. The area of the background should cover any possible character.

**Comment by daschuer** [comment]
OK, I have forced pushed a better solution. m_labelFontSize is now adjusted to be Font Type independent.

File: src/widget/woverview.cpp [hunk 6]
```diff
@@ -955,13 +955,6 @@ void WOverview::drawMarkLabels(QPainter* pPainter, const float offset, const flo
                 continue;
             }
         }
-        if (pMark->m_label.intersects(m_cuePositionLabel)
-                || pMark->m_label.intersects(m_cueTimeDistanceLabel)
-                || pMark->m_label.intersects(m_timeRulerPositionLabel)
-                || pMark->m_label.intersects(m_timeRulerDistanceLabel)) {
-            continue;
-        }
```
Comment (Be-ing) [inline src/widget/woverview.cpp]: Revert this. Labels should not overlap.

File: src/widget/woverview.cpp [hunk 7]
```diff
@@ -107,12 +107,18 @@ void WOverview::setup(const QDomNode& node, const SkinContext& context) {
...
+    if (!labelFontSize) {
+        labelFontSize = 12.0;
     }
 
+    // Since we use the system font here, we need to adjust for full height
+    // of "Üg" and not just high of "Ug" set by setPixelSize
+    QFont refFont;
+    refFont.setPixelSize(labelFontSize);
+    QFontMetricsF fontMetrics(refFont);
+    m_labelFontSize = labelFontSize / fontMetrics.height() * labelFontSize;
```
Comment (Be-ing) [inline src/widget/woverview.cpp]: I don't understand this math. Why not just use `fontMetrics.height()`?

Comment (daschuer) [reply src/widget/woverview.cpp]: The labelFontSize is the high of "Hg" decent() + capitalHight() 
My system font is Ubuntu, you have a different font. The ascent() is different.
Here we adjust the m_labelFontSize for an almost equal font independent hight().

I think I need to improve the comment.

Does this produce a good result on you side?

Comment (Be-ing) [reply src/widget/woverview.cpp]: >  almost equal font independent hight()

This goal does not make sense to me. The actual height of the text as it is rendered is what is needed.

> My system font is Ubuntu, you have a different font.

This seems to be the root of the issues. Does Shade not specify a font? If so, let's fix that issue so we get consistent results on both our computers.

File: src/widget/woverview.cpp (hunk 6 above)
Comment (Be-ing) [reply src/widget/woverview.cpp]: ping

Comment (daschuer) [reply src/widget/woverview.cpp]: It is not the right place to force this here. This code skips to draw the label just because the font is too big or the waveform too small. I agree that the labels should not overlap, but if there is a skin issue it should be rendered anyway to see the overlap.

File: src/widget/woverview.cpp (hunk 7 above)
Comment (daschuer) [reply src/widget/woverview.cpp]: >> almost equal font independent hight()

> This goal does not make sense to me. The actual height of the text as it is rendered is what is needed.

Yes, but the actual height is finally kind of random since the skin designer does not know the system font. The skin designer can optimize the looking with the pixel size of its local font. If user uses a font with the same pixel size but taller or lower characters, the labels might overlap. This has happened on your and my system. This math here takes care, that the actual pixel size for the tall font is lower in a way that Üg = _Üg_ and not only Ug = _Ug_

Comment (daschuer) [reply src/widget/woverview.cpp]: > Does Shade not specify a font?

It does: ```font-family: "Open Sans";```
But as long as we do not install the font it can be any fallback.

Comment (Be-ing) [reply src/widget/woverview.cpp]: But does that style get applied to WOverview? Does it need to be explicitly specified for WOverview?

File: src/widget/woverview.cpp (hunk 6 above)
Comment (Be-ing) [reply src/widget/woverview.cpp]: This is required for normal functionality. It is not to work around bugs in skins. It is normal for the intro/outro cue markers and durations which are rendered on top, and the main cue label which is rendered at vertical center in Tango, to overlap the labels when holding right click.

> if there is a skin issue it should be rendered anyway to see the overlap.

I disagree. The label disappearing makes bugs easier to catch.

File: src/widget/woverview.cpp (hunk 7 above)
Comment (Be-ing) [reply src/widget/woverview.cpp]: > the actual height is finally kind of random since the skin designer does not know the system font

This is the root problem that needs to be addressed. IMO it makes little sense to try to work around this with overcomplicated code.

Comment (daschuer) [reply src/widget/woverview.cpp]: Ok, do you have an alternative solution?

File: src/widget/woverview.cpp (hunk 6 above)
Comment (daschuer) [reply src/widget/woverview.cpp]: Some pixel vertical overlap is no issue, rectifies not drawing the lable, especially because the position and distance labels are still painted overlapping. We need here a solution that detects the cue type and hides the labels based on that.

File: src/widget/woverview.cpp (hunk 7 above)
Comment (daschuer) [reply src/widget/woverview.cpp]: How about setting the maximum label height in the skin xml rather then setting the font pixel size?
This way the font can defined by style sheets and if it exceeds the lable the size is adjusted.

File: src/widget/woverview.cpp (hunk 6 above)
Comment (Be-ing) [reply src/widget/woverview.cpp]: No, we need to not have overlap. Please do not make this any more complicated. It was already working well. I will not merge this without this commit being reverted (or removed with a rebase).

Comment (daschuer) [reply src/widget/woverview.cpp]: We need a solution for the root issue first: Different font heights on different systems. Will you fix it in your branch? 
Than I can finally adjust this code here.

**Comment by Be-ing** [comment]
@ronso0 do you have any hints why we are seeing different results with the labels being hidden on our different computers?

I think if we cannot figure out why this is happening soon we should just stick with a font size of 9 for Shade's labels and get 2.3 beta out already. I don't think increasing the font size by a pixel or two is worth continuing to delay this.

File: src/widget/woverview.cpp (hunk 7 above)
Comment (Be-ing) [reply src/widget/woverview.cpp]: I have tried replacing `font-family: "Open Sans"` with `font-family: Ubuntu` in Shade's style.qss and the label still gets hidden when hovering.

Comment (daschuer) [reply src/widget/woverview.cpp]: The font used in my case is already Ubuntu. We need to find a way to define the Mixxx shipped font on all systems.

Comment (Be-ing) [reply src/widget/woverview.cpp]: With `font-family: Ubuntu`, I am able to increase the LabelFontSize for Shade up to 12. With "Open Sans", 11 is the max size that fits. Either we can decrease the font size to 11 or explicitly specify to use the Ubuntu font for Shade (which we also use for Tango).

Comment (Be-ing) [reply src/widget/woverview.cpp]: Ping. You can either explicitly specify the Ubuntu font or change the font size to 11. I think explicitly specifying the font would be better to ensure a consistent appearance.

**Comment by ronso0** [comment]
> @ronso0 do you have any hints why we are seeing different results with the labels being hidden on our different computers?

Not at first glance. I can have a look at it soonish, like tomorrow night.
Did you make sure the font include is working in other places of Shade? We deliver OpenSans with Mixxx so why should it fall back to some system font..

**Comment by daschuer** [comment]
Looking on this again, I think it is a conceptual issue to set the font size via skin.xml but the font type via css. Can we remove the size feature form the skin?

**Comment by daschuer** [comment]
Ok, WOverview has the cue labels and the status messages, so it makes sense the set the label font size via skin xml. I have reverted the critical commits and set the font to "Open Sans" for WOverview explicit. Finally it looks nice. @Be-ing, ok for merging?

**Comment by Be-ing** [comment]
> With font-family: Ubuntu, I am able to increase the LabelFontSize for Shade up to 12. With "Open Sans", 11 is the max size that fits.

Hmm, I can't reproduce these results now. I'm not sure what I did to get them. The maximum font size with Open Sans is 8 and it is 10 with Ubuntu. I tested this on my branch and got the same results. Without specifying a font-family for WOverview in style.qss, the max font size is 9 as it is in my branch. I do not know what default font that is falling back to.

**Comment by daschuer** [comment]
@Be-ing: Does this mean the issue is fixed for you as well? Merge?

**Comment by Be-ing** [comment]
No, this is still not working.

**Comment by daschuer** [comment]
Why not? Can you attach a screenshot?
Do you have an idea how to fix the issue?

**Comment by Be-ing** [comment]
I stated the fix before: change the font size to 8 with Open Sans or 10 with Ubuntu.

**Comment by daschuer** [comment]
But why? Here it looks nice now. How does it look at you side?

**Comment by Be-ing** [comment]
The problem is as it has been whole time of this PR. The labels disappear when hovered with the mouse if the text is too big because of the overlap.

**Comment by daschuer** [comment]
With my latest changes it doesn't happen any more and it is independent from the system font now.
That's why  screenshot is crucial to compare differences pixel by pixel. 
Does it happen with 100% scaling as well?

**Comment by Be-ing** [comment]
image
image

**Comment by daschuer** [comment]
Oh sorry, I can reproduce. I must have pushed the wrong version. Now I have switched back to Ubuntu 11. 

All skins benefit from the sightly bigger font.

grafik

grafik

grafik

grafik

**Comment by Be-ing** [comment]
The text is still too big. Again, the maximum font size for Ubuntu that fits with Shade is 10. Here is a screenshot of how it looks with this PR currently:
image

**Comment by daschuer** [comment]
Does the 100% version works for you now?

**Comment by daschuer** [comment]
Here is mine 
grafik

**Comment by daschuer** [comment]
and 300 %
grafik

**Comment by daschuer** [comment]
The font has the same size, but the label is bigger. Maybe it depends on the Qt version.

**Comment by daschuer** [comment]
Ups, this is a bigger issue, because the overview deck and the mini decks of every skin are suffering the same issue. 
So the difference seams either be a bug in your or mine Qt version. 
It feels odd to user such big labels, don't allow to overlap, just for the rare case of a very tall character. 
I have tried various exotic characters, but I did not find one which is able to exceed my size. 
I think it is reasonable to reduce the label size in favour of readable numbers. 
 
grafik

**Comment by Be-ing** [comment]
> Does the 100% version works for you now?

No. As before, the scale factor has no impact on this bug.

**Comment by Be-ing** [comment]
> Maybe it depends on the Qt version.

Perhaps? I can't come up with any other explanation for why we are getting different results on our computers.

**Comment by Be-ing** [comment]
> overview deck and the mini decks of every skin are suffering the same issue.

Okay, so decrease the font size.

> I think it is reasonable to reduce the label size in favour of readable numbers.

Again, I don't think we should mess with this. This PR for trivial polishing details has been open for 3 weeks now, IMO it was never necessary in the first place, meanwhile we still have no beta release. Can we please just decrease the font size and move on already?

**Comment by daschuer** [comment]
This PR discovers an other system dependent relationship than we have first discussed. This means the skin designer has no control if the skin works or not. This needs to be fixed before merge.

Reducing the font size for the mimum decks a small value makes the info ugly or even unreadable. With 300 % scaling, you will of course be able to read any font size this is not the case with 100 %.

For a complete solution, I think we need to dig down the issue first. I have already tried to find differences caused by the Qt Version, but reviewing the Qt git history did not discover anything so far. I think I will add some debug statements, to get it finally. 

As a quick hack we may limit the lable high to 1/3 of the overview hight. This will at least stop the "disappearing" bug. Cutting some pixels of some exotic characters is a neglegtable issue compared to this.

What do you think.

**Comment by Be-ing** [comment]
> This needs to be fixed before merge.

No it does not. Please keep perspective here. Nobody expects perfection from a beta release. We are already going on a year without a release. We are talking about literally 1 pixel difference in font sizes.

>  What do you think.

No more hacks. Just reduce the font size and let's get this release out already.

**Comment by daschuer** [comment]
> We are talking about literally 1 pixel difference in font sizes.

That would be no issue for me with the Shade main overview. The real issue is the unreliable label background size and the unreadable font size in the overview and preview player. I cannot fix this here, because the label too big issue on your side.

**Comment by Be-ing** [comment]
How about we introduce a boolean skin element to disable showing the times on cue hover for the preview deck and mini decks?

**Comment by daschuer** [comment]
Good idea. So we need to find only the bug for the different label heights.

**Comment by Be-ing** [comment]
No we don't. Just make the text one pixel size smaller.

**Comment by daschuer** [comment]
OK

**Comment by Be-ing** [comment]
If you want to look into the differences in rendering later, you can do that. But continuing to hold this up over literally one pixel in font size is IMO absurd.
